### PR TITLE
[ENG-1662] feat: Example refactor of smartlead

### PR DESCRIPTION
**Pull request by RajatPawar**
This is a draft PR that builds on top of the work @Cobalt0s has done to ease understanding & implementation & improve reusability / accessibility for new connectors. This is no way complete, and exists as an example (ignore the linter errors!)

### CODE REVIEW

**Comment by RajatPawar** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1207?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1207** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1207?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1207?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1205** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1205?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: connector/new.go [hunk 1]
```diff
@@ -216,9 +216,11 @@ func newPipelinerConnector(parameters Parameters) (connectors.Connector, error)
...
 
 func newSmartleadConnector(parameters Parameters) (connectors.Connector, error) {
-	return smartlead.NewConnector(
-		smartlead.WithAuthenticatedClient(parameters.AuthenticatedClient),
-	)
+	// Client could directly be set via common methods -
+	// hc, _ = common.NewHeaderAuthHTTPClient(context.TODO())
+	// parameters.AuthenticatedClient = hc
+
+	return smartlead.NewConnector(parameters)
```
Comment (Cobalt0s) [inline connector/new.go:223]: I reviewed the other PR but will leave a review here.

The options design pattern that was previously used:
( :cry: ) created duplicated files with methods for each connector
( :cry: ) complicated deep connector architecture redesign, where everything had to resolve around this
( :smile: ) provided a strict list of parameters, the code tells what the connector needs.

Unified parameters across connectors (this PR):
( :cry: ) doesn't tell by autocomplete which properties are important and which are irrelevant (ex: connector may not have metadata), when initializing connector it is not clear what properties I must provide (validation errors would guide in this case)
( :neutral_face: ) creates a central place for parameter definition. From the surface connectors look similar so it makes sense. I think there is no expectation for it to rapidly change so it is neither positive nor negative.
( :smile: ) generalizaing deep connector architecture will be easier, there is no need to use go generics by every place that depends on parameters.

**Review by Cobalt0s — COMMENTED** [review]
I left some thoughts as I was reading both PRs and code locally.

File: internal/components/components.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	provider providers.Provider,
+	params connector.Parameters,
+	setup Setup[T],
+) (conn *T, err error) {
+	defer goutils.PanicRecovery(func(cause error) {
+		err = cause
+		conn = nil
+	})
+
+	clientComponent, err := NewClientComponent(provider, params)
```
Comment (jlimatampersand) [inline internal/components/components.go:26]: I think this function should look something like

```
providerCompoenent, err := newProviderComponent(...)
...
clientComponent, err := newClientComponent(...)
...
connectorComponent, err := newConnectorComponent(...)

return initializeWithComponents(providerComponenent, clientComponent, connectorComponent)

```

File: internal/components/metadata.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import (
+	"context"
+
+	"github.com/amp-labs/connectors/common"
+)
+
+// MetadataStrategy describes an object's schema / metadata.
+type MetadataStrategy interface {
+	// Run accepts an object, and returns a field map for it.
+	Run(ctx context.Context, object string) (*common.ObjectMetadata, error)
```
[the remark this answers is not in the document]
Comment (jlimatampersand) [reply internal/components/metadata.go:12]: Making the interface to have only 1 method `Run` make it impossible for me to understand what this interface is about.

**Review by jlimatampersand — CHANGES_REQUESTED** [review]
Left some comments. There are various moving parts here that makes me to hard to understand the entire picture on how we want to refactor our connectors. 

I think the key here is that we want to have each provider will have their own set of some configurations, then we can import them in some pattern of

```
// function names are not exact

info, err := providerInfo(providername)

info.DoSomething()
.
.
.
.

```

```
// salesforce.go
var config = &ProviderSpecificConfig{
  ...,
  ...,
  ...,

}
```

then NewConnector functions would import `salesforce.ProviderSalesforceConfig` and use it

Overall, I see some patterns can benefit us, whereas other patterns, I am not too convinced. I think it would be the best if we can actually have a session to go over these patterns and @RajatPawar  you can explain why these were implemented such ways. Without explanation, it is hard to understand and it probably will be the same for most contributors.

File: internal/components/components.go (hunk 2 above)
Comment (Cobalt0s) [reply internal/components/components.go:26]: I think that indeed we will end up doing this the explicit style. This is equivalent to DI technique I used in `uber-dig`, but instead it "removed" these lines by making them implicit, which made things un intuitive.

In general any connector constructor could be initialized using some core components which would be configured by getting optional nested components: WithComponentX, WithComponentY. This would make the dependency tree as flat as possible with max of 2 layers and capture the differences among deep connector implementation.

File: internal/components/client.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	cc := &ClientComponent{
+		JSON: &common.JSONHTTPClient{
+			HTTPClient: &common.HTTPClient{
+				Client: params.AuthenticatedClient,
+				ErrorHandler: interpreter.ErrorHandler{
+					JSON: interpreter.NewFaultyResponder(errorFormats(), statusCodeMapping()),
```
[the remark this answers is not in the document]
Comment (Cobalt0s) [reply internal/components/client.go:30]: I am not sure how to go with `errorFormats()` method, because it is suggesting a [Template Pattern](https://refactoring.guru/design-patterns/template-method) which is hard to achieve without inheritence. This behaviour was realized using composition and DI, but as you saw in my PRs it is not easy. Also the desire to use inheritence does suggest that the issue is not easy to resolve.

`someErrorCatalog(provider)()` suggests a one place for all errors for each connector creating yet another switch/map and central place to modify deep connectors.

Instance of error handler should be an argument to `NewClientComponent` method, and that should be designed as a dedicated PR and initiative, it has its own scope and also relates to creating a consistent Ampersand error contract for all deep connectors and bridging that with provider errors.

File: connector/new.go [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+
+func NewConnector(provider providers.Provider, parameters Parameters) (connectors.Connector, error) {
+	var (
+		conn connectors.Connector
+		err  error
+	)
+
+	switch provider {
+	case providers.Hubspot:
+		conn, err = buildConnector(newHubspotConnector, parameters, mustAuthenticatedClient)
```
[the remark this answers is not in the document]
Comment (Cobalt0s) [reply connector/new.go:43]: @RajatPawar as an alternative that we discussed the other day we could try interface pair per parameter.
See this PR https://github.com/amp-labs/connectors/pull/1069. This pattern was used to infer catalog variables just from the `parameters` struct alone without any required deep connector code.

And to combine on the optional connector approach that Jay is suggesting here the deep connector code would be either or:
```
type parameters struct {
	paramsbuilder.ClientRequired
	paramsbuilder.WorkspaceRequired
	paramsbuilder.ModuleRequired
}

type parameters struct {
	paramsbuilder.ClientRequired
	paramsbuilder.WorkspaceRequired
	paramsbuilder.ModuleOptional
}
```

And the internal receptor package would get a passed `parameters` instance and there all the manipulation would happen. It would ask the following: is parameters a client holder? is it a workspace holder? then is it optional? Those if statements would allow to make all the necessary jugment to conclude if the struct is well formed.

In other words, this existing method:
```
func (p parameters) ValidateParams() error {
	return errors.Join(
		p.Client.ValidateParams(),
		p.Workspace.ValidateParams(),
		p.Module.ValidateParams(),
	)
}
```
will be encoded into `paramsbuilder.ModuleOptional`, etc.

We could try and spike a `parameters PR`. My main goal is, parameters struct declration holds all the information, weight, essence.

**Comment by RajatPawar** [comment]
Creating a new PR for an easier reviewing experience :)
